### PR TITLE
feat(liquidator) Re-try Logic

### PR DESCRIPTION
**Pull request by chrismaree**
Implement re-try logic in the liquidator bot.

**Motivation**

Right now the bots fall over immediately with any kind of error. This can be due to a flaky API or other issues that are not serious in nature but simply require a retry. Often, this requires the bot to be re-booted to solve the issue. Having the bots implement their own re-try logic will drastically reduce the DevOps burden in these cases.

**Summary**

Implement re-try logic within the execution loop in the liquidator bot. This implementation uses the [async-retry](https://github.com/vercel/async-retry) package.

**Issue(s)**

First implementation for issue #1812. Subsequent PRs will implement the same logic for the disputer & monitor bots.

### CODE REVIEW

File: liquidator/liquidator.js [hunk 1]
```diff
@@ -125,14 +125,12 @@ class Liquidator {
     });
 
     // If an override is provided, use that price. Else, get the latest price from the price feed.
-    const price = liquidatorOverridePrice ? this.toBN(liquidatorOverridePrice) : this.priceFeed.getCurrentPrice();
+    const price = liquidatorOverridePrice
+      ? this.toBN(liquidatorOverridePrice.toString())
```
Comment (chrismaree) [inline liquidator/liquidator.js:129]: unrelated change: I added this `.toString` as this would throw if the param was a number. this is bad behavior as it would crash the bot if set to a non-string.

File: liquidator/liquidator.js [hunk 2]
```diff
@@ -125,14 +125,12 @@ class Liquidator {
...
+      : this.priceFeed.getCurrentPrice();
 
     if (!price) {
-      this.logger.warn({
-        at: "Liquidator",
-        message: "Cannot liquidate: price feed returned invalid value"
-      });
-      return;
+      throw new Error("Cannot liquidate: price feed returned invalid value");
```
Comment (chrismaree) [inline liquidator/liquidator.js:133]: Also, if there is no price the bot needs to fail more loudly than a `warn`. A side effect of this is I can test the re-try logic with invalid price feeds :)

File: liquidator/test/Liquidator.js [hunk 3]
```diff
@@ -171,13 +171,16 @@ contract("Liquidator.js", function(accounts) {
     assert.equal((await emp.getCollateral(sponsor1)).rawValue, toWei("125"));
     assert.equal((await emp.getCollateral(sponsor2)).rawValue, toWei("150"));
 
-    // No liquidations if the price feed returns an invalid value.
+    // Liquidator throws an error if the price feed returns an invalid value.
```
Comment (chrismaree) [inline liquidator/test/Liquidator.js:174]: needed to update this test to accommodate the `throw` on null price feed.

File: liquidator/liquidator.js (hunk 2 above)
Comment (chrismaree) [reply liquidator/liquidator.js:133]: curious to see if you guys agree with this change.

File: liquidator/index.js [hunk 4]
```diff
@@ -128,22 +139,43 @@ async function run(logger, empAddress, pollingDelay, priceFeedConfig, liquidator
...
+          await liquidator.liquidatePositions(currentSyntheticBalance, liquidatorOverridePrice);
+          // Check for any finished liquidations that can be withdrawn.
+          await liquidator.withdrawRewards();
+        },
+        {
+          retries: errorRetries,
+          onRetry: error => {
+            logger.debug({
+              at: "Liquidator#index",
+              message: "An error was thrown in the execution loop - retrying",
```
Comment (nicholaspai) [inline liquidator/index.js:160]: Maybe add here which retry out of max retry attempts you're on?

File: liquidator/index.js [hunk 5]
```diff
@@ -128,22 +139,43 @@ async function run(logger, empAddress, pollingDelay, priceFeedConfig, liquidator
...
+              error: typeof error === "string" ? new Error(error) : error
+            });
+          }
+        }
+      );
       // If the polling delay is set to 0 then the script will terminate the bot after one full run.
       if (pollingDelay === 0) {
+        logger.debug({
+          at: "Liquidator#index",
+          message: "End of severless execution loop - yielding process"
```
Comment (nicholaspai) [inline liquidator/index.js:170]:
```suggestion
          message: "End of serverless execution loop - yielding process"
```

Comment (nicholaspai) [reply liquidator/index.js:170]: Also, is `yielding` intended to mean `stopping`?

File: liquidator/liquidator.js (hunk 2 above)
Comment (nicholaspai) [reply liquidator/liquidator.js:133]: I'm good with this, the bot can't really do anything without a price.

File: liquidator/index.js (hunk 4 above)
Comment (chrismaree) [reply liquidator/index.js:160]: I wanted to do that but it's not available in the this function, unfortunately. you can access it in the re-try loop but not in the `onRetry` call.

File: liquidator/liquidator.js [hunk 6]
```diff
@@ -418,7 +416,7 @@ class Liquidator {
       let resolvedPrice;
       if (requestTimestamp) {
         try {
-          resolvedPrice = revertWrapper(
+          resolvedPrice = resolvedPrice = revertWrapper(
```
Comment (nicholaspai) [inline liquidator/liquidator.js:419]: typo? I see `resolvedPrice` twice

File: liquidator/index.js (hunk 5 above)
Comment (chrismaree) [reply liquidator/index.js]: ye, can change the wording to something else if we want.

File: liquidator/liquidator.js (hunk 6 above)
Comment (chrismaree) [reply liquidator/liquidator.js:419]: not sure how that happened; have fixed

File: liquidator/index.js (hunk 4 above)
Comment (nicholaspai) [reply liquidator/index.js:160]: huh ok I think its actually helpful enough to figure out a way to add it, perhaps we can pass into the `onRetry` callback somehow. WDYT @UMAprotocol/eng

Comment (mrice32) [reply liquidator/index.js:160]: It's a little messy, but I think you could do something like:
```js
while(true) {
  let numRetries = 0;
  await retry(async () => {
    numRetires++;
    ...
  },
  {
    ...
    message: `retry ${numRetries} failed with error ${error}`
  });
}
```

**Comment by mrice32** [comment]
> This is really cool thanks for implementing! I'm curious why you decided not to set some sort of timeout between retries? I don't think we need to neccessarily, but I noticed thats a feature in the `async-retry` lib.

This is only a thought (I haven't tested), but if we ran into an infura websocket connection drop, a small (parameterizable) delay may help give the websocket provider time to re-establish the connection before retrying the run. Not entirely sure if that's how that works tho.

**Comment by daywiss** [comment]
looks good. i have some reservations about adding the retry library as a dep, but i think given the design it probably makes it easier to reason about. curious to see how the infura errors get handled in this solution.

**Comment by chrismaree** [comment]
> This all looks great to me!
> 
> I think we should probably just reach consensus on:
> 
> 1. Should we have a delay between retries?
> 2. (less important) should we messily implement a way to spit out the retry number in the log?

Cool. So in terms of your questions:
1. The default right now is 1 second between retries. do we want this to be longer?
2. Regarding the messy re-try counter, I don't really think it's needed and makes the implementation a bit messy.

**Comment by mrice32** [comment]
> > This all looks great to me!
> > I think we should probably just reach consensus on:
> > 
> > 1. Should we have a delay between retries?
> > 2. (less important) should we messily implement a way to spit out the retry number in the log?
> 
> Cool. So in terms of your questions:
> 
> 1. The default right now is 1 second between retries. do we want this to be longer?
> 2. Regarding the messy re-try counter, I don't really think it's needed and makes the implementation a bit messy.

1. Why not a little longer, like 15 seconds or however long our usual node connection timeout is?
2. SGTM. @nicholaspai is that cool with you?

**Comment by chrismaree** [comment]
> > > This all looks great to me!
> > > I think we should probably just reach consensus on:
> > > 
> > > 1. Should we have a delay between retries?
> > > 2. (less important) should we messily implement a way to spit out the retry number in the log?
> > 
> > 
> > Cool. So in terms of your questions:
> > 
> > 1. The default right now is 1 second between retries. do we want this to be longer?
> > 2. Regarding the messy re-try counter, I don't really think it's needed and makes the implementation a bit messy.
> 
> 1. Why not a little longer, like 15 seconds or however long our usual node connection timeout is?
> 2. SGTM. @nicholaspai is that cool with you?

Cool. I set this to the same value that the truffle config uses (10 seconds).

**Comment by nicholaspai** [comment]
> > > This all looks great to me!
> > > I think we should probably just reach consensus on:
> > > 
> > > 1. Should we have a delay between retries?
> > > 2. (less important) should we messily implement a way to spit out the retry number in the log?
> > 
> > 
> > Cool. So in terms of your questions:
> > 
> > 1. The default right now is 1 second between retries. do we want this to be longer?
> > 2. Regarding the messy re-try counter, I don't really think it's needed and makes the implementation a bit messy.
> 
> 1. Why not a little longer, like 15 seconds or however long our usual node connection timeout is?
> 2. SGTM. @nicholaspai is that cool with you?

SGTM on point 2, also agreed that a bit of delay between retries makes sense and provides more room for error
